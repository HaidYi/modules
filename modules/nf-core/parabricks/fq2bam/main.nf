process PARABRICKS_FQ2BAM {
    tag "$meta.id"
    label 'process_high'
    label 'process_gpu'
    label 'gpu'
    stageInMode 'copy'

    container "nvcr.io/nvidia/clara/clara-parabricks:4.4.0-1"

    input:
    tuple val(meta), val(read_group), path ( r1_fastq, stageAs: "?/*"), path ( r2_fastq, stageAs: "?/*")
    tuple path(fasta), path(fai)
    tuple val(meta1), path(index)
    path known_sites
    path interval_file

    output:
    tuple val(meta), path("*.bam"), path("*.bai") , emit: bam_bai
    tuple val(meta), path("qc_metrics/*"), optional: true, emit: qc_metrics
    tuple val(meta), path("*.table"), optional: true, emit: bqsr_table
    tuple val(meta), path("*.duplicate-metrics.txt"), optional: true, emit: duplicate_metrics
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "Parabricks module does not support Conda. Please use Docker / Singularity / Podman instead."
    }

    def args = task.ext.args ?: ''
    def prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"    
    def known_sites_command = known_sites ? (known_sites instanceof List ? known_sites.collect { "--knownSites $it" }.join(' ') : "--knownSites ${known_sites}") : ""
    def known_sites_output = known_sites ? "--out-recal-file ${prefix}.table" : ""
    def interval_file_command = interval_file ? (interval_file instanceof List ? interval_file.collect { "--interval-file $it" }.join(' ') : "--interval-file ${interval_file}") : ""
    def num_gpus = task.accelerator ? "--num-gpus $task.accelerator.request" : ''

    def readgroups_string = read_group.collect { rg -> "@RG\\tID:${rg.read_group}__${rg.sample}\\tSM:${rg.sample}\\tPL:${rg.platform}\\tLB:${rg.sample}\\tPU:${rg.read_group}" }

    def in_fq_command = meta.single_end 
        ? (r1_fastq instanceof List 
            ? r1_fastq.collect { "--in-se-fq $it" }.join(' ') 
            : "--in-se-fq ${r1_fastq}"
        )
        : (r1_fastq instanceof List && r2_fastq instanceof List && readgroups_string instanceof List
            ? (r1_fastq.indexed().collect { idx, r1 -> "--in-fq $r1 ${r2_fastq[idx]} \"${readgroups_string[idx]}\"" }).join(' ')
            : "--in-fq ${r1_fastq} ${r2_fastq} \"${readgroups_string.join(' ')}\""
        )
 
    """
    INDEX=`find -L ./ -name "*.amb" | sed 's/\\.amb\$//'`
    cp $fasta \$INDEX
    
    pbrun \\
        fq2bam \\
        --ref \$INDEX \\
        $in_fq_command \\
        --read-group-sm $meta.id \\
        --out-bam ${prefix}.bam \\
        $num_gpus \\
        $known_sites_command \\
        $known_sites_output \\
        $interval_file_command \\
        --out-qc-metrics-dir qc_metrics \\
        --out-duplicate-metrics ${prefix}.duplicate-metrics.txt \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
            pbrun: \$(echo \$(pbrun version 2>&1) | sed 's/^Please.* //' )
    END_VERSIONS
    """

    stub:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "Parabricks module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    
    def args = task.ext.args ?: ''
    def prefix     = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"    
    def known_sites_command = known_sites ? (known_sites instanceof List ? known_sites.collect { "--knownSites $it" }.join(' ') : "--knownSites ${known_sites}") : ""
    def known_sites_output = known_sites ? "--out-recal-file ${prefix}.table" : ""
    def interval_file_command = interval_file ? (interval_file instanceof List ? interval_file.collect { "--interval-file $it" }.join(' ') : "--interval-file ${interval_file}") : ""

    def readgroups_string = read_group.collect { rg -> "@RG\\tID:${rg.read_group}__${rg.sample}\\tSM:${rg.sample}\\tPL:${rg.platform}\\tLB:${rg.sample}\\tPU:${rg.read_group}" }

    def in_fq_command = meta.single_end 
        ? (r1_fastq instanceof List 
            ? r1_fastq.collect { "--in-se-fq $it" }.join(' ') 
            : "--in-se-fq ${r1_fastq}"
        )
        : (r1_fastq instanceof List && r2_fastq instanceof List && readgroups_string instanceof List
            ? (r1_fastq.indexed().collect { idx, r1 -> "--in-fq $r1 ${r2_fastq[idx]} \"${readgroups_string[idx]}\"" }).join(' ')
            : "--in-fq ${r1_fastq} ${r2_fastq} \"${readgroups_string.join(' ')}\""
        )

    def metrics_output_command = args = "--out-duplicate-metrics duplicate-metrics.txt" ? "touch duplicate-metrics.txt" : ""
    def known_sites_output_command = known_sites ? "touch ${prefix}.table" : ""
    def qc_metrics_output_command = args = "--out-qc-metrics-dir qc_metrics " ? "mkdir qc_metrics && touch qc_metrics/alignment.txt" : ""
    """

    echo $in_fq_command

    touch run.log
    touch ${prefix}.bam
    touch ${prefix}.bam.bai
    $metrics_output_command
    $known_sites_output_command
    $qc_metrics_output_command
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
            pbrun: \$(echo \$(pbrun version 2>&1) | sed 's/^Please.* //' )
    END_VERSIONS
    """
}
