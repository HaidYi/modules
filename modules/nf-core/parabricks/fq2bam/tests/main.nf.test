nextflow_process {

    name "Test Process PARABRICKS_FQ2BAM"
    script "../main.nf"
    process "PARABRICKS_FQ2BAM"

    tag "bwa/index"
    tag "modules"
    tag "parabricks/fq2bam"
    tag "modules_nfcore"
    tag "parabricks"
    tag "gpu"

    setup {
        run("BWA_INDEX") {
            script "../../../bwa/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true)
                ])
                """
            }
        }

        run("BWA_INDEX", alias: 'BWA_INDEX_PE') {
            script "../../../bwa/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("SRR389222 - fastq - se") {

        config './nextflow.config'

        when {
            params {
                module_args = '--low-memory'
                // Ref: https://forums.developer.nvidia.com/t/problem-with-gpu/256825/6
                // Parabricks’s fq2bam requires 24GB of memory.
                // Using --low-memory for testing
            }
            process {
                """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':true ],
                    [ [ 'read_group':'rg', 'sample':'sm', 'platform':'pl' ] ],
				    [
				        file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/SRR389222_sub1.fastq.gz', checkIfExists: true)
				    ],
                    [],
                    []
				])
                input[1] = Channel.of([
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true),
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam_bai[0][1]).getReadsMD5(),
                    file(process.out.bam_bai[0][2]).name,
                    process.out.versions,
                    path(process.out.versions[0]).yaml
                ).match() }
            )
        }
    }

    test("SRR389222 - fastq - se - stub") {
        options '-stub'

        when {
            process {
               """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':true ],
                    [ [ 'read_group':'rg', 'sample':'sm', 'platform':'pl' ] ],
				    [
				        file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/SRR389222_sub1.fastq.gz', checkIfExists: true)
				    ],
                    [],
                    []
				])
                input[1] = Channel.of([
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true),
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml
                    ).match() }
            )
        }
    }

    test("sarscov2 - fastq - pe") {

        config './nextflow.config'

        when {
            params {
                module_args = '--low-memory'
                // Ref: https://forums.developer.nvidia.com/t/problem-with-gpu/256825/6
                // Parabricks’s fq2bam requires 24GB of memory.
                // Using --low-memory for testing
            }
            process {
                """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':false ],
                    [ [ 'read_group':'rg', 'sample':'sm', 'platform':'pl' ] ],
				    [
				        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
				    ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
				    ],
                    []
				])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam_bai[0][1]).getReadsMD5(),
                    file(process.out.bam_bai[0][2]).name,
                    process.out.versions,
                    path(process.out.versions[0]).yaml
                ).match() }
            )
        }

    }

    test("sarscov2 - fastq - pe - stub") {

        options '-stub'

        when {
            process {
                """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':false ],
                    [ [ 'read_group':'rg', 'sample':'sm', 'platform':'pl' ] ],
				    [
				        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
				    ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
				    ],
                    []
				])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml
                    ).match() }
            )
        }

    }

    test("sarscov2 - fastq - pe - merge") {

        config './nextflow.config'

        when {
            params {
                module_args = '--low-memory'
                // Ref: https://forums.developer.nvidia.com/t/problem-with-gpu/256825/6
                // Parabricks’s fq2bam requires 24GB of memory.
                // Using --low-memory for testing
            }
            process {
                """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':false ],
                    [ [ 'read_group':'rg1', 'sample':'sm1', 'platform':'pl' ], [ 'read_group':'rg2', 'sample':'sm1', 'platform':'pl' ] ],
				    [
				        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
				    ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
				    ],
                    []
				])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam_bai[0][1]).getReadsMD5(),
                    file(process.out.bam_bai[0][2]).name,
                    process.out.versions,
                    path(process.out.versions[0]).yaml
                ).match() }
            )
        }

    }

    test("sarscov2 - fastq - pe - merge - stub") {

        options '-stub'

        when {
            params {
                module_args = '--low-memory'
                // Ref: https://forums.developer.nvidia.com/t/problem-with-gpu/256825/6
                // Parabricks’s fq2bam requires 24GB of memory.
                // Using --low-memory for testing
            }
            process {
                """
                input[0] = Channel.of([
				    [ 'id':'test', 'read_group':'rg', 'sample':'sm', 'platform':'pl', 'single_end':false ],
                    [ [ 'read_group':'rg1', 'sample':'sm1', 'platform':'pl' ], [ 'read_group':'rg2', 'sample':'sm1', 'platform':'pl' ] ],
				    [
				        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
				    ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
				    ],
                    []
				])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
				input[2] = BWA_INDEX.out.index
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam_bai[0][1]).getReadsMD5(),
                    file(process.out.bam_bai[0][2]).name,
                    process.out.versions,
                    path(process.out.versions[0]).yaml
                ).match() }
            )
        }

    }

}