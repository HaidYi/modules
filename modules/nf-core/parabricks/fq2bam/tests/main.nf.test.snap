{
    "SRR389222 - fastq - se - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": true
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": true
                        },
                        "alignment.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
                ],
                "bam_bai": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": true
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bqsr_table": [
                    
                ],
                "duplicate_metrics": [
                    
                ],
                "qc_metrics": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": true
                        },
                        "alignment.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
                ]
            },
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-01-23T17:54:47.091148531"
    },
    "sarscov2 - fastq - pe - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": false
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": false
                        },
                        "alignment.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
                ],
                "bam_bai": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": false
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bqsr_table": [
                    
                ],
                "duplicate_metrics": [
                    
                ],
                "qc_metrics": [
                    [
                        {
                            "id": "test",
                            "read_group": "rg",
                            "sample": "sm",
                            "platform": "pl",
                            "single_end": false
                        },
                        "alignment.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
                ]
            },
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-01-23T18:12:27.155758204"
    },
    "sarscov2 - fastq - pe - merge": {
        "content": [
            "6a2854a2742873d280ca03af1e8567d5",
            "test.bam.bai",
            [
                "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
            ],
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-01-23T18:13:15.174625228"
    },
    "sarscov2 - fastq - pe": {
        "content": [
            "50a751c2aaaaeaa8c341d948aaa9cc0d",
            "test.bam.bai",
            [
                "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
            ],
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-01-23T18:12:19.009705387"
    },
    "SRR389222 - fastq - se": {
        "content": [
            "3d5b94990c7fdf90a682edb5ee0f59de",
            "test.bam.bai",
            [
                "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
            ],
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2024-12-16T12:24:32.45197929"
    },
    "sarscov2 - fastq - pe - merge - stub": {
        "content": [
            "d41d8cd98f00b204e9800998ecf8427e",
            "test.bam.bai",
            [
                "versions.yml:md5,55d1e67ef8fa9d0ea3065363a653ffef"
            ],
            {
                "PARABRICKS_FQ2BAM": {
                    "pbrun": "4.4.0-1"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-01-23T18:10:03.788922476"
    }
}